object-shorthand rule does not properly handle computed properties

**What version of ESLint are you using?**

2.8.0

**What parser (default, Babel-ESLint, etc.) are you using?**

Default

**Please show your full configuration:**

```json
{
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module"
  },
  "rules": {
    "object-shorthand": ["error", "always"]
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
var foo = {
  bar: function() {},
  ['bar']: function() {},
  [bar]: function() {},

  baz: baz,
  ['baz']: baz
};
```

**What did you expect to happen?**

I would have expected all properties in this object to have errors.

**What actually happened? Please include the actual, raw output from ESLint.**

```
  2:3  error  Expected method shorthand    object-shorthand
  6:3  error  Expected property shorthand  object-shorthand
```

When using computed properties, object shorthand was not checked (admittedly, the string literal computed properties are a pretty dumb case, since they could just be used directly as keys).

I'm happy to take a stab at this one if someone can confirm this is not intentional.

--- a/docs/rules/object-shorthand.md
+++ b/docs/rules/object-shorthand.md
@@ -49,8 +49,9 @@ Each of the following properties would warn:
 /*eslint-env es6*/
 
 var foo = {
-    x: function() {},
-    y: function *() {},
+    w: function() {},
+    x: function *() {},
+    [y]: function() {},
     z: z
 };
 ```
@@ -62,8 +63,9 @@ In that case the expected syntax would have been:
 /*eslint-env es6*/
 
 var foo = {
-    x() {},
-    *y() {},
+    w() {},
+    *x() {},
+    [y]() {},
     z
 };
 ```

--- a/lib/rules/object-shorthand.js
+++ b/lib/rules/object-shorthand.js
@@ -108,8 +108,13 @@ module.exports = {
                     return;
                 }
 
-                // getters, setters and computed properties are ignored
-                if (node.kind === "get" || node.kind === "set" || node.computed) {
+                // only computed methods can fail the following checks
+                if (!APPLY_TO_METHODS && node.computed) {
+                    return;
+                }
+
+                // getters and setters are ignored
+                if (node.kind === "get" || node.kind === "set") {
                     return;
                 }
 

--- a/tests/lib/rules/object-shorthand.js
+++ b/tests/lib/rules/object-shorthand.js
@@ -65,6 +65,10 @@ ruleTester.run("object-shorthand", rule, {
         { code: "var x = {['y']: 'y'}", parserOptions: { ecmaVersion: 6 }, options: ["properties"] },
         { code: "var x = {['y']: y}", parserOptions: { ecmaVersion: 6 }, options: ["properties"] },
 
+        // object literal computed methods
+        { code: "var x = {[y]() {}}", parserOptions: { ecmaVersion: 6 }, options: ["methods"] },
+        { code: "var x = {[y]: function x() {}}", parserOptions: { ecmaVersion: 6 }, options: ["methods"] },
+
         // options
         { code: "var x = {y() {}}", parserOptions: { ecmaVersion: 6 }, options: ["methods"] },
         { code: "var x = {x, y() {}, a:b}", parserOptions: { ecmaVersion: 6 }, options: ["methods"] },
@@ -99,11 +103,14 @@ ruleTester.run("object-shorthand", rule, {
         { code: "doSomething({'x': x})", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Expected property shorthand.", type: "Property" }] },
         { code: "doSomething({a: 'a', 'x': x})", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Expected property shorthand.", type: "Property" }] },
         { code: "doSomething({y: function() {}})", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Expected method shorthand.", type: "Property" }] },
+        { code: "doSomething({[y]: function() {}})", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Expected method shorthand.", type: "Property" }] },
+        { code: "doSomething({['y']: function() {}})", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Expected method shorthand.", type: "Property" }] },
 
         // options
         { code: "var x = {y: function() {}}", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Expected method shorthand.", type: "Property" }], options: ["methods"] },
         { code: "var x = {x, y() {}, z: function() {}}", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Expected method shorthand.", type: "Property" }], options: ["methods"] },
         { code: "var x = {ConstructorFunction: function(){}, a: b}", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Expected method shorthand.", type: "Property" }], options: ["methods"] },
+        { code: "var x = {[y]: function() {}}", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Expected method shorthand.", type: "Property" }], options: ["methods"] },
         { code: "var x = {x: x}", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Expected property shorthand.", type: "Property" }], options: ["properties"] },
         { code: "var x = {a, b, c(){}, x: x}", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Expected property shorthand.", type: "Property" }], options: ["properties"] },
         { code: "var x = {y() {}}", parserOptions: { ecmaVersion: 6 }, errors: [{ message: "Expected longform method syntax.", type: "Property" }], options: ["never"] },

